File: apps/cli-py/src/betteruseofai/saving.py
"""What using smaller models has saved, mirroring packages/core/src/saving.ts.

A counterfactual with a stated baseline: every turn re-priced as if it had gone
to rung zero of its own family, minus what it actually cost. A turn already on
the frontier model saves nothing. A turn on an unrecognised model cannot be
re-priced and is left out, and the count of those is returned so the page can
say so. Every figure stays a range and both estimates share rows, region and
scope, so the difference is between like and like.
"""

from __future__ import annotations

from dataclasses import dataclass, replace
from typing import Any

from .engine import Range, get_model
from .estimate import Estimate, UsageEvent, estimate

SAVING_BASELINE = "the largest model in each family"


@dataclass
class DaySaving:
    day: str
    energy_wh: Range | None


@dataclass
class Saving:
    baseline: str
    energy_wh: Range | None
    water_ml: Range | None
    carbon_g: Range | None
    by_day: list[DaySaving]
    skipped: int


def _plus(a: Range | None, b: Range | None) -> Range | None:
    if a is None:
        return b
    if b is None:
        return a
    return Range(a.low + b.low, a.central + b.central, a.high + b.high)


def _minus(a: Range | None, b: Range | None) -> Range | None:
    """A minus B, floored at nothing. A saving is never negative."""
    if a is None or b is None:
        return None
    return Range(
        max(0.0, a.low - b.high), max(0.0, a.central - b.central), max(0.0, a.high - b.low)
    )


def saving(
    pairs: list[tuple[UsageEvent, Estimate]],
    dataset: dict[str, Any],
    region_code: str | None = None,
    water_scope: str = "on-site + off-site",
    carbon_basis: str = "location-based",
) -> Saving:
    energy: Range | None = None
    water: Range | None = None
    carbon: Range | None = None
    skipped = 0
    days: dict[str, Range | None] = {}

    for event, actual in pairs:
        day = event.timestamp[:10]
        model = get_model(actual.model_id, dataset)
        if model is None or actual.energy_wh is None:
            skipped += 1
            continue
        if model["ordinal"] == 0:
            # Already the top of the ladder. Nothing saved, nothing skipped.
            days.setdefault(day, None)
            continue
        top = next(
            (m for m in dataset["models"] if m["family"] == model["family"] and m["ordinal"] == 0),
            None,
        )
        if top is None:
            skipped += 1
            continue

        counterfactual = estimate(
            replace(event, model_raw=top["id"]),
            dataset,
            region_code=region_code,
            water_scope=water_scope,
            carbon_basis=carbon_basis,
        )

        saved = _minus(counterfactual.energy_wh, actual.energy_wh)
        energy = _plus(energy, saved)
        water = _plus(water, _minus(counterfactual.water_ml, actual.water_ml))
        carbon = _plus(carbon, _minus(counterfactual.carbon_g, actual.carbon_g))
        days[day] = _plus(days.get(day), saved)

    return Saving(
        baseline=SAVING_BASELINE,
        energy_wh=energy,
        water_ml=water,
        carbon_g=carbon,
        by_day=[DaySaving(day, days[day]) for day in sorted(days)],
        skipped=skipped,
    )

File: packages/core/src/saving.ts
import { estimate } from './estimate.js';
import type { CarbonBasis, Dataset, Estimate, Range, UsageEvent, WaterScope } from './types.js';

/**
 * What using smaller models has saved, against the frontier model in each
 * family.
 *
 * Nothing here can know what somebody would have done without the tool. What
 * it can compute honestly is a counterfactual with a stated baseline: every
 * turn re-priced as if it had gone to rung zero of its own family, minus what
 * it actually cost. A turn that was already on the frontier model saves
 * nothing. A turn on an unrecognised model cannot be re-priced and is left out,
 * and the count of those is returned so the page can say so.
 *
 * Every figure stays a range, and the two estimates share the same rows,
 * region and scope, so the difference is between like and like. This started
 * life in the extension; it moved here so the dashboard the command line tools
 * write draws the same rings from the same arithmetic, and so the Python tool
 * has one definition to mirror.
 */

export interface DaySaving {
  day: string;
  energyWh: Range | null;
}

export interface Saving {
  baseline: string;
  energyWh: Range | null;
  waterMl: Range | null;
  carbonG: Range | null;
  byDay: DaySaving[];
  /** Turns that could not be re-priced, so are not in the figures. */
  skipped: number;
}

export interface SavingInput {
  event: UsageEvent;
  estimate: Estimate;
}

export interface SavingOptions {
  regionCode?: string;
  waterScope?: WaterScope;
  carbonBasis?: CarbonBasis;
}

export const SAVING_BASELINE = 'the largest model in each family';

const plus = (a: Range | null, b: Range | null): Range | null => {
  if (!a) return b;
  if (!b) return a;
  return { low: a.low + b.low, central: a.central + b.central, high: a.high + b.high };
};

/** A minus B, floored at nothing. A saving is never negative. */
const minus = (a: Range | null, b: Range | null): Range | null => {
  if (!a || !b) return null;
  return {
    low: Math.max(0, a.low - b.high),
    central: Math.max(0, a.central - b.central),
    high: Math.max(0, a.high - b.low),
  };
};

export const saving = (inputs: SavingInput[], dataset: Dataset, options: SavingOptions = {}): Saving => {
  let energy: Range | null = null;
  let water: Range | null = null;
  let carbon: Range | null = null;
  let skipped = 0;
  const days = new Map<string, Range | null>();

  for (const { event, estimate: actual } of inputs) {
    const day = event.timestamp.slice(0, 10);
    const model = actual.modelId ? dataset.models.find((one) => one.id === actual.modelId) : undefined;
    if (!model || !actual.energyWh) {
      skipped += 1;
      continue;
    }
    if (model.ordinal === 0) {
      // Already the top of the ladder. Nothing saved, nothing skipped.
      if (!days.has(day)) days.set(day, null);
      continue;
    }
    const top = dataset.models.find((one) => one.family === model.family && one.ordinal === 0);
    if (!top) {
      skipped += 1;
      continue;
    }

    const counterfactual = estimate({ ...event, modelId: top.id, modelRaw: top.id }, { dataset, ...options });

    const saved = minus(counterfactual.energyWh, actual.energyWh);
    energy = plus(energy, saved);
    water = plus(water, minus(counterfactual.waterMl, actual.waterMl));
    carbon = plus(carbon, minus(counterfactual.carbonG, actual.carbonG));
    days.set(day, plus(days.get(day) ?? null, saved));
  }

  return {
    baseline: SAVING_BASELINE,
    energyWh: energy,
    waterMl: water,
    carbonG: carbon,
    byDay: [...days.entries()]
      .sort((a, b) => (a[0] < b[0] ? -1 : 1))
      .map(([day, energyWh]) => ({ day, energyWh })),
    skipped,
  };
};
